src/profiler.ts

```diff
@@ -13,57 +13,292 @@
+   * on the type of profile created, an error will be thrown.
+   */
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent'));
+    return this.pollProfilerService();
+  }
+
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   *
+   * The profiler server will be polled for instructions at most once every
+   * minTimeBetweenProfilesMillis.
+   */
+  async pollProfilerService(): Promise<void> {
+    const startCreateMillis = Date.now();
+    const prof = await this.createProfile();
+    await this.profileAndUpload(prof);
+    const endCreateMillis = Date.now();
+
+    // Schedule the next profile.
+    const delayMillis = Math.max(
+        0,
+        this.config.minTimeBetweenProfilesMillis -
+            (endCreateMillis - startCreateMillis));
+
+    setTimeout(this.pollProfilerService.bind(this), delayMillis).unref();
+  }
+
+  /**
+   * Talks to Stackdriver Profiler server, which hangs until server indicates
+   * job should be profiled.
+   *
+   * If any problem is encountered, the problem will be logged and
+   * createProfile() will be retried.
+   *
+   * TODO: implement backoff and retry when error encountered. createProfile()
+   * should be retried at time response indicates this request should be retried
+   * or with exponential backoff (up to one hour) if the response does not
+   * indicate when to retry this request. Once this is implemented, an error
+   * will be thrown only if the error indicates one definitely should not
+   * retry createProfile.
+   *
+   * Public to allow for testing.
+   */
+  async createProfile(): Promise<RequestProfile> {
+    const reqBody = {
+      deployment: {
+        projectId: this.config.projectId,
+        target: this.config.serviceContext.service,
+        labels: {zone: this.config.zone, instance: this.config.instance},
+      },
+      profileType: this.profileTypes,
+    };
+    const options = {
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+
+    let requestError: Error|undefined = undefined;
+    let retryRequest = true;
+    try {
+      const [body, response] = await this.request(options);
+      if (isErrorResponseCode(response.statusCode)) {
+        retryRequest = isRetriableResponseCode(response.statusCode);
+        requestError =
+            new Error('Error creating profile: ' + response.statusMessage);
+        this.logger.debug(requestError);
+      }
+      return body;
+    } catch (err) {
+      retryRequest = isRetriableError(err);
+      requestError = new Error('Error creating profile: ' + err);
+      this.logger.debug(requestError);
+    }
+    if (retryRequest) {
+      // TODO: check response to see if response specifies a backoff.
+      // TODO: implement exponential backoff.
+      await delay(this.config.minTimeBetweenProfilesMillis);
```

Comment:
I understand we don't have the exponential backoff yet, that could be a separate PR, but please name that variable like something with "backoff" in its name. Look at the Go profiling agent code maybe.